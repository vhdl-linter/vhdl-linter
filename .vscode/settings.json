{
  "typescript.tsdk": "node_modules/typescript/lib",
  "eslint.lintTask.enable": true,
  "eslint.lintTask.options": "--ext .ts .",
  "eslint.format.enable": true,
  "cSpell.words": [
    "actuals",
    "clientrc",
    "downto",
    "Inoutput",
    "Mentionables",
    "nand",
    "Onig",
    "OSVVM",
    "quickfix",
    "Referencable",
    "setq",
    "subarchitecture",
    "Textmate",
    "ulogic",
    "uninstantiated",
    "Uniquify",
    "Verilog",
    "vhdl",
    "Vhdl",
    "vmode",
    "vprop",
    "vunit",
    "Whitespaces",
    "xnor"
  ]
}