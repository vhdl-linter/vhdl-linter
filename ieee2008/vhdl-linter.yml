rules:
  type-resolved: false
  unit: false