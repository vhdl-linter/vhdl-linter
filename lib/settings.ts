import { DeepPartial } from 'utility-types';
export interface ISettings {
  ports: {
    outRegex: string;
    inRegex: string;
    enablePortStyle: boolean;
  };
  paths: {
    additional: string[];
    ignoreRegex: string;
  };
  style: {
    preferredLogicTypePort: "unresolved" |  "resolved" | "ignore";
    preferredLogicTypeSignal: "unresolved" | "resolved" | "ignore";
    unusedSignalRegex: string;
    ieeeCasing: 'lowercase' | 'UPPERCASE';
  };
  rules: {
    warnLibrary: boolean;
    warnLogicType: boolean;
    warnMultipleDriver: boolean;
  };
  semanticTokens: boolean;
}
export const defaultSettings: ISettings = {
  ports: {
    outRegex: '^o_',
    inRegex: '^i_',
    enablePortStyle: true,
  },
  paths: {
    additional: [],
    ignoreRegex: ''
  },
  style: {
    preferredLogicTypePort: 'unresolved',
    preferredLogicTypeSignal: 'unresolved',
    unusedSignalRegex: '_unused$',
    ieeeCasing: 'lowercase'
  },
  rules: {
    warnLogicType: true,
    warnLibrary: false,
    warnMultipleDriver: false
  },
  semanticTokens: false
};
export function defaultSettingsGetter() {
  return defaultSettings;
}
export function defaultSettingsWithOverwrite(overwrite?: DeepPartial<ISettings>) {
  const newDefault = Object.assign({}, defaultSettings);

  if (overwrite) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    recursiveObjectAssign(newDefault as any, overwrite);
  }
  return () => defaultSettings;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function recursiveObjectAssign<T extends Record<string, any>>(target: T, source: Partial<T>) {
  Object.keys(source).forEach(key => {
    const s_val = source[key];
    const t_val = target[key];
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (target as any)[key] = t_val && s_val && typeof t_val === 'object' && typeof s_val === 'object'
      ? recursiveObjectAssign(t_val, s_val)
      : s_val;
  });
  return target;
}