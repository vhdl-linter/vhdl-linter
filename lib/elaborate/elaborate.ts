import { DiagnosticSeverity } from "vscode-languageserver/node";
import { implementsIHasReference, implementsIHasDefinitions } from "../parser/interfaces";
import { OArchitecture, OEntity, OFile, OPackage, OPackageBody } from "../parser/objects";
import { VhdlLinter } from "../vhdl-linter";
import { elaborateAliases } from "./elaborate-aliases";
import { elaborateAssociations } from "./elaborate-association";
import { elaborateComponents } from "./elaborate-components";
import { elaborateInstantiations } from "./elaborate-instantiations";
import { ElaborateReferences } from "./elaborate-references";
import { ElaborateSelectedNames } from "./elaborate-selected-names";
import { elaborateUseClauses } from "./elaborate-use-clauses";

export class Elaborate {
  file: OFile;
  private constructor(public vhdlLinter: VhdlLinter) {
    this.file = vhdlLinter.file;
  }
  public static async elaborate(vhdlLinter: VhdlLinter) {

    const elaborator = new Elaborate(vhdlLinter);
    await elaborator.elaborateAll();
  }
  public static clear(vhdlLinter: VhdlLinter) {
    for (const obj of vhdlLinter.file.objectList) {
      if (implementsIHasReference(obj)) {
        obj.referenceLinks = [];
        obj.aliasReferences = [];
      }
      if (implementsIHasDefinitions(obj)) {
        obj.definitions = [];
      }
      if (obj instanceof OEntity) {
        obj.correspondingArchitectures = [];
      } else if (obj instanceof OArchitecture) {
        obj.correspondingEntity = undefined;
      } else if (obj instanceof OPackageBody) {
        obj.correspondingPackage = undefined;
      } else if (obj instanceof OPackage) {
        obj.correspondingPackageBodies = [];
      }

    }
  }
  async elaborateAll() {

    await this.vhdlLinter.handleCanceled();

    // const start = Date.now();
    // Map architectures to entity
    for (const architecture of this.file.architectures) {
      if (architecture.entityName === undefined) {
        continue;
      }
      // Find entity first in this file
      let entity = this.file.entities.find(entity => entity.lexerToken.getLText() === architecture.entityName.getLText());
      if (!entity) { // Find entity in all files
        entity = this.vhdlLinter.projectParser.entities.find(entity => entity.lexerToken.getLText() === architecture.entityName.getLText());
      }
      if (entity) {
        architecture.correspondingEntity = entity;
      }

    }
    for (const entity of this.file.entities) {
      entity.correspondingArchitectures = this.vhdlLinter.projectParser.architectures.filter(architecture => entity.lexerToken.getLText() === architecture.entityName.getLText());
    }

    // Map package body to package
    for (const pkg of this.file.packages) {
      if (pkg instanceof OPackageBody) {

        // Find entity first in this file
        let pkgHeader: OPackage | undefined = this.file.packages.find(pkgHeader => pkgHeader instanceof OPackage && pkgHeader.lexerToken.getLText() === pkg.lexerToken.getLText()) as OPackage | undefined;
        if (!pkgHeader) { // Find entity in all files
          pkgHeader = this.vhdlLinter.projectParser.packages.find(pkgHeader => pkgHeader instanceof OPackage && pkgHeader.lexerToken.getLText() === pkg.lexerToken.getLText()) ;
        }
        if (pkgHeader) {
          pkg.correspondingPackage = pkgHeader;
          pkgHeader.correspondingPackageBodies.push(pkg);
        } else {
          this.vhdlLinter.addMessage({
            range: pkg.lexerToken.range,
            severity: DiagnosticSeverity.Warning,
            message: `Can not find package for package body.`
          }, 'elaborate');
        }

      }
    }
    //     console.log(packages);
    elaborateUseClauses(this.file, this.vhdlLinter.projectParser, this.vhdlLinter);
    await this.vhdlLinter.handleCanceled();
    //     console.log(packages);
    // elaborateReferences(this.file);


    // console.log(`elaboration: reads for: ${Date.now() - start} ms.`);
    // start = Date.now();
    await this.vhdlLinter.handleCanceled();
    elaborateComponents(this.file, this.vhdlLinter.projectParser);
    await this.vhdlLinter.handleCanceled();
    elaborateInstantiations(this.file, this.vhdlLinter.projectParser);

    // console.log(`elaboration: instantiations for: ${Date.now() - start} ms.`);
    // start = Date.now();
    await this.vhdlLinter.handleCanceled();

    // console.log(`elaboration: components for: ${Date.now() - start} ms.`);
    // start = Date.now();
    await this.vhdlLinter.handleCanceled();
    ElaborateReferences.elaborate(this.vhdlLinter);
    await this.vhdlLinter.handleCanceled();
    elaborateAssociations(this.file);
    await this.vhdlLinter.handleCanceled();
    elaborateAliases(this.file);
    await this.vhdlLinter.handleCanceled();
    ElaborateSelectedNames.elaborate(this.vhdlLinter);
    // console.log(`elaboration: associations for: ${Date.now() - start} ms.`);
    // start = Date.now();
    await this.vhdlLinter.handleCanceled();

  }







}