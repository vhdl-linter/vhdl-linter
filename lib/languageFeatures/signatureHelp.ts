import { MarkupKind, Position, SignatureHelp, SignatureInformation } from "vscode-languageserver";
import { implementsIHasGenerics } from "../parser/interfaces";
import { OAliasWithSignature, OGenericAssociationList } from "../parser/objects";
import { VhdlLinter } from "../vhdl-linter";
import { findParentInstantiation } from "./helper/findParentInstantiation";
export function signatureHelp(linter: VhdlLinter, position: Position): SignatureHelp | null {
  const result = findParentInstantiation(linter, position);
  if (!result) {
    return null;
  }
  const [instantiation, associationList] = result;
  const signatures: SignatureInformation[] = [];
  for (const definition of instantiation.definitions) {
    if (definition instanceof OAliasWithSignature) {
      // Handle AliasWIthSignatures
    } else {
      const portOrGeneric = associationList instanceof OGenericAssociationList && implementsIHasGenerics(definition) ? definition.generics : definition.ports;
      if (portOrGeneric.length === 0) {
        signatures.push({
          label: ''
        });
      } else {
        // Skip definitions with less ports
        if (portOrGeneric.length < (associationList?.children.length ?? 0)) {
          continue;
        }
        const text = portOrGeneric.map(port => port.lexerToken.text).join(', ');
        let activeParameter;
        if (associationList) {
          // Find active parameter
          // If in range of association via number
          const posI = linter.getIFromPosition(position);
          const associationIndex = associationList.children.findIndex(association => association.range.start.i <= posI && association.range.end.i >= posI);
          const association = associationList.children[associationIndex];
          if (associationIndex > -1 && association && association.formalPart.length > 0) {
            for (const formal of association.formalPart) {
              for (const [portIndex, port] of portOrGeneric.entries()) {
                if (port.lexerToken.getLText() === formal.referenceToken.getLText()) {
                  activeParameter = portIndex;
                }
              }
            }
            // If not found set outside of range.
            // LSP standard is not complete on this.
            // This might be hacky, client maybe supposed to reset to zero anyways. But works in vscode
            if (activeParameter === undefined) {
              activeParameter = portOrGeneric.length;
            }
          } else {
            activeParameter = 0;
            for (const [childNumber, child] of associationList.children.entries()) {
              // Extend the end range by white spaces (assume it belongs to the association if cursor is in whitespace)
              let tokenIndex = linter.file.lexerTokens.findIndex(token => token.range.end.i === child.range.end.i);
              while (linter.file.lexerTokens[tokenIndex + 1]?.isWhitespace()) {
                tokenIndex++;
              }
              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
              if (posI >= (linter.file.lexerTokens[tokenIndex]!.range.end.i)) {
                activeParameter = childNumber + 1;
              }
            }

          }
        }

        signatures.push({
          label: text,
          parameters: portOrGeneric.map(port => ({
            label: port.lexerToken.text,
            documentation: {
              kind: MarkupKind.Markdown,
              value: '```vhdl\n' + port.range.getText().replaceAll(/\s+/g, ' ') + '\n```'
            }
          })),
          activeParameter
        });
      }

    }
  }


  return {
    signatures
  };
}