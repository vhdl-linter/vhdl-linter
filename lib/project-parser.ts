import { FSWatcher, watch } from 'chokidar';
import { EventEmitter } from 'events';
import { promises } from 'fs';
import { join, sep } from 'path';
import { cwd } from 'process';
import { OContext, OEntity, OPackage, OPackageInstantiation } from './parser/objects';
import { SettingsGetter, VhdlLinter } from './vhdl-linter';

export class ProjectParser {

  public cachedFiles: FileCache[] = [];
  public packages: OPackage[] = [];
  public packageInstantiations: OPackageInstantiation[] = [];
  public contexts: OContext[] = [];
  public entities: OEntity[] = [];
  events = new EventEmitter();
  private watchers: FSWatcher[] = [];
  // Constructor can not be async. So constructor is private and use factory to create
  public static async create(workspaces: string[], fileIgnoreRegex: string, settingsGetter: SettingsGetter, disableWatching = false) {
    const projectParser = new ProjectParser(workspaces, fileIgnoreRegex, settingsGetter);
    await projectParser.init(disableWatching);
    return projectParser;
  }
  private constructor(public workspaces: string[], public fileIgnoreRegex: string, public settingsGetter: SettingsGetter) { }
  public addFolders(pathes: string[]) {
    for (const path of pathes) {
      const watcher = watch(path.replace(sep, '/') + '/**/*.vhd?(l)', { ignoreInitial: true });
      watcher.on('add', async (path) => {
        const cachedFile = await (FileCache.create(path, this));
        this.cachedFiles.push(cachedFile);
        this.flattenProject();
        this.events.emit('change', 'add', path);
      });
      watcher.on('change', async (path) => {
        // console.log('change', path);
        const cachedFile = process.platform === 'win32'
          ? this.cachedFiles.find(cachedFile => cachedFile.path.toLowerCase() === path.toLowerCase())
          : this.cachedFiles.find(cachedFile => cachedFile.path === path);
        if (cachedFile) {
          await cachedFile.parse();
          this.flattenProject();
          this.events.emit('change', 'change', path);
        } else {
          console.error('modified file not found', path);
        }
      });
      watcher.on('unlink', path => {
        const cachedFileIndex = this.cachedFiles.findIndex(cachedFile => cachedFile.path === path);
        if (cachedFileIndex > -1) {
          this.cachedFiles.splice(cachedFileIndex, 1);
          this.flattenProject();
          this.events.emit('change', 'unlink', path);
        }
      });
      this.watchers.push(watcher);
    }
  }
  private async init(disableWatching: boolean) {
    const files = new Set<string>();
    await Promise.all(this.workspaces.map(async (directory) => {
      const directories = await this.parseDirectory(directory);
      return (await Promise.all(directories.map(file => promises.realpath(file)))).forEach(file => files.add(file));
    }));
    (await this.parseDirectory(join(cwd(), 'ieee2008'))).forEach(file => files.add(file));

    for (const file of files) {
      const cachedFile = await FileCache.create(file, this);
      this.cachedFiles.push(cachedFile);
    }
    this.cachedFiles.sort((a, b) => b.lintingTime - a.lintingTime);
    // console.log('Times: \n' + this.cachedFiles.slice(0, 10).map(file => `${file.path}: ${file.lintingTime}ms`).join('\n'));
    this.flattenProject();
    if (!disableWatching) {
      this.addFolders(this.workspaces);
    }
  }
  async stop() {
    for (const watcher of this.watchers) {
      await watcher.close();
    }
  }
  private async parseDirectory(directory: string): Promise<string[]> {
    const files: string[] = [];
    const entries = await promises.readdir(directory);
    const ignoreRegex = this.fileIgnoreRegex.trim().length > 0 ? new RegExp(this.fileIgnoreRegex) : null;

    // const entries = await promisify(directory.getEntries)()
    await Promise.all(entries.map(async entry => {
      try {
        const filePath = directory + sep + entry;
        const fileStat = await promises.stat(filePath);
        if (fileStat.isFile()) {
          if (entry.match(/\.vhdl?$/i) && (ignoreRegex === null || !filePath.match(ignoreRegex))) {
            files.push(filePath);
          }
        } else {
          files.push(... await this.parseDirectory(filePath));
        }
      } catch (e) {
        console.error(e);
      }
    }));
    return files;
  }
  private flattenProject() {
    this.packages = [];
    this.packageInstantiations = [];
    this.entities = [];
    for (const cachedFile of this.cachedFiles) {
      this.entities.push(...cachedFile.entities);
      if (cachedFile.packages) {
        this.packages.push(...cachedFile.packages);
      }
      if (cachedFile.packageInstantiations) {
        this.packageInstantiations.push(...cachedFile.packageInstantiations);
      }
      if (cachedFile.contexts) {
        this.contexts.push(...cachedFile.contexts);
      }
    }
  }

}

export class FileCache {
  packages?: OPackage[];
  packageInstantiations?: OPackageInstantiation[];
  contexts: OContext[] = [];
  entities: OEntity[] = [];
  text: string;
  linter: VhdlLinter;
  lintingTime: number;
  // Constructor can not be async. So constructor is private and use factory to create
  public static async create(path: string, projectParser: ProjectParser) {
    const cache = new FileCache(path, projectParser);
    await cache.parse();
    return cache;
  }
  private constructor(public path: string, public projectParser: ProjectParser) {
  }
  async parse() {
    this.text = await promises.readFile(this.path, { encoding: 'utf8' });
    this.linter = new VhdlLinter(this.path, this.text, this.projectParser, this.projectParser.settingsGetter);
    this.packages = this.linter.file.packages.filter((p): p is OPackage => p instanceof OPackage);
    this.packageInstantiations = this.linter.file.packageInstantiations;
    this.entities = this.linter.file.entities;
    this.contexts = this.linter.file.contexts;
  }

}