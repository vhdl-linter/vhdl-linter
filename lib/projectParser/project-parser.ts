import { FSWatcher, watch } from 'chokidar';
import { EventEmitter } from 'events';
import { promises, readFileSync } from 'fs';
import { Worker } from 'node:worker_threads';
import { join, sep } from 'path';
import { OContext, OEntity, OFile, OPackage, OPackageBody } from '../parser/objects';
import * as objects from '../parser/objects';
import { VhdlLinter } from '../vhdl-linter';
// import { StaticPool } from 'node-worker-threads-pool';
import { Parser } from '../parser/parser';
import { pool } from './parser-pool';
export class ProjectParser {

  public cachedFiles: OFileCache[] = [];
  private packages: (OPackage | OPackageBody)[];
  private contexts: OContext[] = [];
  private entities: OEntity[];
  events = new EventEmitter();
  constructor(public workspaces: string[], public fileIgnoreRegex: string) { }
  async init() {
    const files = new Set<string>();
    await Promise.all(this.workspaces.map(async (directory) => {
      // console.log('dir', directory);
      const directories = await this.parseDirectory(directory);
      return (await Promise.all(directories.map(file => promises.realpath(file)))).forEach(file => files.add(file));
    }));
    // for (const directory of this.workspaces) {
    //   this.parseDirectory(directory).forEach(file => files.add(realpathSync(file)));
    // }
    const pkg = __dirname;
    if (pkg) {
      //       console.log(pkg, new Directory(pkg + '/ieee2008'));
      (await this.parseDirectory(join(pkg, `${sep}..${sep}..${sep}..${sep}ieee2008`))).forEach(file => files.add(file));
    }
    for (const file of files) {
      const cachedFile = new OFileCache(file);
      await cachedFile.parse();
      this.cachedFiles.push(cachedFile);
    }
    // this.cachedFiles.push(...await Promise.all(Array.from(files).map(async file => {
    //   const cachedFile = new OFileCache(file, this);
    //   await cachedFile.parse();
    //   return cachedFile;
    // })));
    this.cachedFiles.sort((a, b) => b.lintingTime - a.lintingTime);
    // console.log('Times: \n' + this.cachedFiles.slice(0, 10).map(file => `${file.path}: ${file.lintingTime}ms`).join('\n'));
    this.fetchEntitesAndPackagesAndContexts();
    for (const workspace of this.workspaces) {
      const watcher = watch(workspace.replace(sep, '/') + '/**/*.vhd', { ignoreInitial: true });
      watcher.on('add', async (path) => {
        const cachedFile = new OFileCache(path);
        await cachedFile.parse();
        this.cachedFiles.push(cachedFile);
        this.fetchEntitesAndPackagesAndContexts();
        this.events.emit('change', 'add', path);
      });
      watcher.on('change', async (path) => {
        // console.log('change', path);
        const cachedFile = process.platform === 'win32'
          ? this.cachedFiles.find(cachedFile => cachedFile.path.toLowerCase() === path.toLowerCase())
          : this.cachedFiles.find(cachedFile => cachedFile.path === path);
        if (cachedFile) {
          await cachedFile.parse();
        } else {
          console.error('modified file not found', path);
        }
        this.fetchEntitesAndPackagesAndContexts();
        this.events.emit('change', 'change', path);
      });
      watcher.on('unlink', path => {
        const cachedFileIndex = this.cachedFiles.findIndex(cachedFile => cachedFile.path === path);
        this.cachedFiles.splice(cachedFileIndex, 1);
        this.fetchEntitesAndPackagesAndContexts();
        this.events.emit('change', 'unlink', path);
      });

    }
  }
  private async parseDirectory(directory: string): Promise<string[]> {
    const files: string[] = [];
    const entries = await promises.readdir(directory);
    const ignoreRegex = this.fileIgnoreRegex.trim().length > 0 ? new RegExp(this.fileIgnoreRegex) : null;

    // const entries = await promisify(directory.getEntries)()
    await Promise.all(entries.map(async entry => {
      try {
        const filePath = directory + sep + entry;
        const fileStat = await promises.stat(filePath);
        if (fileStat.isFile()) {
          if (entry.match(/\.vhdl?$/i) && (ignoreRegex === null || !filePath.match(ignoreRegex))) {
            files.push(filePath);
          }
        } else {
          files.push(... await this.parseDirectory(filePath));
        }
      } catch (e) {
        console.error(e);
      }
    }));
    return files;
  }
  private fetchEntitesAndPackagesAndContexts() {
    //     console.log(this.cachedFiles);
    this.packages = [];
    this.entities = [];
    for (const cachedFile of this.cachedFiles) {
      this.entities.push(...cachedFile.entities);
      if (cachedFile.packages) {
        this.packages.push(...cachedFile.packages);
      }
      if (cachedFile.contexts) {
        this.contexts.push(...cachedFile.contexts);
      }
    }
  }
  watcher: FSWatcher;
  addFolders(folders: string[]) {
    this.watcher.add(folders.map(folder => folder.replace(sep, '/') + '/**/*.vhd'));
  }
  public getPackages() {
    return this.packages;
  }
  public getContexts() {
    return this.contexts;
  }
  public getEntities() {
    return this.entities;
  }
}

export class OFileCache {

  path: string;
  digest: string;
  file: OFile;
  packages?: (OPackage | OPackageBody)[];
  contexts: OContext[] = [];
  entities: OEntity[] = [];
  lintingTime: number;
  constructor(file: string) {
    // this.digest = await file.getDigest();
    this.path = file;
  }
  async parse() {
    console.log(this.path, 'start');
    // const parser = new Parser(readFileSync(this.path, { encoding: 'utf-8' }), this.path, true);
    // this.file = parser.parse();

    this.file = await pool.exec({
      path: this.path,
    });
    console.log(this.path, 'end');
    this.packages = this.file.packages;
    this.entities = this.file.entities;
    this.contexts = this.file.contexts;
  }
}