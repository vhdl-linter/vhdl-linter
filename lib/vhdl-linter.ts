import {
  CancellationToken,
  CodeAction, CodeActionKind, Diagnostic, DiagnosticSeverity, LSPErrorCodes, Position, Range, ResponseError, TextEdit
} from 'vscode-languageserver';
import { Elaborate } from './elaborate/elaborate';
import { FileParser } from './parser/file-parser';
import {
  OFile, OI, OIRange, ParserError
} from './parser/objects';
import { ProjectParser } from './project-parser';
import { rules } from './rules/rule-index';
import { ISettings } from './settings';

export interface IAddSignalCommandArguments {
  textDocumentUri: string;
  signalName: string;
  position: OI;
}
export interface OIDiagnostic extends Diagnostic {
  range: OIRange;
}
export interface IIgnoreLineCommandArguments {
  textDocumentUri: string;
  range: Range;
}
export type SettingsGetter = (resource: URL) => Promise<ISettings> | ISettings;
type diagnosticCodeActionCallback = (textDocumentUri: string) => CodeAction[];
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export class VhdlLinter {
  messages: Diagnostic[] = [];
  file: OFile;
  parser: FileParser;
  parsedSuccessfully = false;
  constructor(public uri: URL, public text: string, public projectParser: ProjectParser,
    public settingsGetter: SettingsGetter,
    public token?: CancellationToken) {
    try {
      this.parser = new FileParser(text, this.uri);
      this.file = this.parser.parse();
      this.parsedSuccessfully = true;
      this.file.parserMessages = this.parser.state.messages;
    } catch (e) {
      if (e instanceof ParserError) {
        const solution = e.solution;
        let code;
        if (solution) {
          code = this.addCodeActionCallback((textDocumentUri: string) => {
            const actions = [];
            actions.push(CodeAction.create(
              solution.message,
              {
                changes: {
                  [textDocumentUri]: solution.edits
                }
              },
              CodeActionKind.QuickFix));
            return actions;
          });
        }
        // Include the parser messages that did not result in fatal.
        // Normally this is done in RParser. (This rule can not be called here as it needs settings which need to be fetched async and this is constructor)
        this.messages.push(...this.parser.state.messages);
        this.messages.push({
          range: e.range,
          severity: DiagnosticSeverity.Error,
          message: e.message,
          code
        });
        this.file = new OFile(this.text, this.uri, this.text, this.parser.lexerTokens);
      } else {
        let message = 'Unknown error while parsing';
        if (e instanceof Error) {
          message = `Javascript error while parsing '${e.message}' ${e.stack ?? ''}`;
        }
        this.messages.push({
          range: Range.create(Position.create(0, 0), Position.create(10, 10)),
          message
        });
        console.error(e);
        this.file = new OFile(this.text, this.uri, this.text, []);

      }
    }
    //     console.log(`done parsing: ${editorPath}`);

  }

  diagnosticCodeActionRegistry: diagnosticCodeActionCallback[] = [];
  addCodeActionCallback(handler: diagnosticCodeActionCallback): number {
    return this.diagnosticCodeActionRegistry.push(handler) - 1;
  }

  addMessage(diagnostic: OIDiagnostic, name: string): void {
    if (this.checkMagicComments(diagnostic.range, name)) {
      const newCode = this.addCodeActionCallback((textDocumentUri: string) => {
        const actions: CodeAction[] = [];
        actions.push(CodeAction.create(
          `Ignore ${name} on this line.`,
          {
            changes: {
              [textDocumentUri]: [
                TextEdit.insert(Position.create(diagnostic.range.end.line, 1000), ` -- vhdl-linter-disable-line ${name}`)]
            }
          },
          CodeActionKind.QuickFix));
        actions.push(CodeAction.create(
          `Ignore ${name} for this file.`,
          {
            changes: {
              [textDocumentUri]: [
                TextEdit.insert(Position.create(0, 0), `-- vhdl-linter-disable ${name}\n`)]
            }
          },
          CodeActionKind.QuickFix));
        return actions;
      });
      const codes = [];
      if (typeof diagnostic.code !== 'undefined') {
        codes.push(diagnostic.code);
      }
      codes.push(newCode);
      diagnostic.code = codes.join(';');
      diagnostic.message = diagnostic.message + ` (${name})`;
      this.messages.push(diagnostic);
    }
  }

  private checkMagicComments(range: OIRange, name: string) {
    const matchingMagiComments = this.file.magicComments.filter(magicComment => {
      if (range.start.i < magicComment.range.start.i) {
        return false;
      }
      if (range.end.i > magicComment.range.end.i) {
        return false;
      }
      return true;
    }).filter(magicComment => {
      if (magicComment.rule) {
        return name === magicComment.rule;
      }
      return true;
    });
    return matchingMagiComments.length === 0;


  }



  async handleCanceled() {
    await new Promise(resolve => setImmediate(resolve));
    if (this.token?.isCancellationRequested) {

      throw new ResponseError(LSPErrorCodes.RequestCancelled, 'canceled');
    }
  }
  async checkAll(profiling = false) {
    if (this.parsedSuccessfully === false) {
      return this.messages;
    }
    if (profiling) {
      console.profile();
    }
    let start;
    let i = 0;
    start = Date.now();
    try {
      await Elaborate.elaborate(this);
      if (profiling) {
        console.log(`check ${i++}: ${Date.now() - start}ms`);
        start = Date.now();
      }
      // await this.removeBrokenActuals();
      if (profiling) {
        console.log(`check ${i++}: ${Date.now() - start}ms`);
        start = Date.now();
      }
      await this.handleCanceled();

      const settings = await this.settingsGetter(this.uri);
      for (const checkerClass of rules) {
        if ((settings.rules as Record<string, boolean>)[checkerClass.ruleName]) {
          const checker = new checkerClass(this, settings);
          checker.check();
          if (profiling) {
            console.log(`check ${checkerClass.ruleName}: ${Date.now() - start}ms`);
            start = Date.now();
          }
        }
        await this.handleCanceled();
      }
    } catch (err) {
      if ((err instanceof ResponseError && err.code === LSPErrorCodes.RequestCancelled)) {
        throw err;
      } else if (err instanceof ParserError) {
        this.messages.push(Diagnostic.create(err.range, `Error while parsing: '${err.message}'`));
      } else {
        this.messages.push(Diagnostic.create(Range.create(Position.create(0, 0), Position.create(10, 100)), `Error while checking: '${err.message}'\n${err.stack ?? ''}`));
      }
    }

    if (profiling) {
      console.profileEnd();
    }
    return this.messages;
  }

  getIFromPosition(p: Position): number {
    const text = this.text.split('\n').slice(0, p.line);
    const i = text.join('\n').length + p.character;
    return i;
  }
}
