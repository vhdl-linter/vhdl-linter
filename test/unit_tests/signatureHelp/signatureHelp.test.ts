import { afterAll, beforeAll, expect, test } from '@jest/globals';
import { readFileSync } from 'fs';
import { join } from 'path';
import { SignatureHelp } from 'vscode';
import { Position } from 'vscode-languageserver';
import { Elaborate } from '../../../lib/elaborate/elaborate';
import { signatureHelp } from '../../../lib/languageFeatures/signatureHelp';
import { ProjectParser } from '../../../lib/project-parser';
import { defaultSettingsGetter } from '../../../lib/settings';
import { VhdlLinter } from '../../../lib/vhdl-linter';

let projectParser: ProjectParser;
beforeAll(async () => {
  projectParser = await ProjectParser.create([__dirname], '', defaultSettingsGetter);
});
afterAll(async () => {
  await projectParser.stop();
});
async function prepare(fileName: string) {
  const path = join(__dirname, fileName);
  const linter = new VhdlLinter(path, readFileSync(path, { encoding: 'utf8' }), projectParser, defaultSettingsGetter);
  await Elaborate.elaborate(linter);
  return linter;
}
test('Signature help snapshot port', async () => {
  const linter = await prepare('entity.vhd');
  const help = await signatureHelp(linter, Position.create(7, 6));

  expect(help).toMatchInlineSnapshot(`
{
  "signatures": [
    {
      "activeParameter": 0,
      "label": "port1, port2, port3",
      "parameters": [
        {
          "documentation": {
            "kind": "markdown",
            "value": "\`\`\`vhdl
port1 : in integer
\`\`\`",
          },
          "label": "port1",
        },
        {
          "documentation": {
            "kind": "markdown",
            "value": "\`\`\`vhdl
port2 : in integer
\`\`\`",
          },
          "label": "port2",
        },
        {
          "documentation": {
            "kind": "markdown",
            "value": "\`\`\`vhdl
port3 : in integer
\`\`\`",
          },
          "label": "port3",
        },
      ],
    },
  ],
}
`);

});
test('Signature help snapshot generic', async () => {
  const linter = await prepare('entity.vhd');
  const help = await signatureHelp(linter, Position.create(18, 14));

  expect(help).toMatchInlineSnapshot(`
{
  "signatures": [
    {
      "activeParameter": 0,
      "label": "GENERIC_A, GENERIC_B",
      "parameters": [
        {
          "documentation": {
            "kind": "markdown",
            "value": "\`\`\`vhdl
GENERIC_A : integer
\`\`\`",
          },
          "label": "GENERIC_A",
        },
        {
          "documentation": {
            "kind": "markdown",
            "value": "\`\`\`vhdl
GENERIC_B : integer
\`\`\`",
          },
          "label": "GENERIC_B",
        },
      ],
    },
  ],
}
`);

});
test('Signature help on non existent formal', async () => {
  const linter = await prepare('entity.vhd');
  const help = await signatureHelp(linter, Position.create(15, 8));
  expect(help?.signatures).toHaveLength(1);

  expect((help as SignatureHelp).signatures[0].activeParameter).toBe(3);
});
test('Signature help snapshot active parameter', async () => {
  const linter = await prepare('entity.vhd');
  expect(signatureHelp(linter, Position.create(8, 8))?.signatures[0].activeParameter).toBe(2);
  expect(signatureHelp(linter, Position.create(9, 11))?.signatures[0].activeParameter).toBe(1);
});
test('Signature help snapshot active parameter for partially filled', async () => {
  const linter = await prepare('component.vhd');
  expect(signatureHelp(linter, Position.create(23, 13))?.signatures[0].activeParameter).toBe(0);
  // Comma entered before and after
  expect(signatureHelp(linter, Position.create(26, 12))?.signatures[0].activeParameter).toBe(0);
  expect(signatureHelp(linter, Position.create(26, 13))?.signatures[0].activeParameter).toBe(1);
  // only part of formal
  expect(signatureHelp(linter, Position.create(30, 12))?.signatures[0].activeParameter).toBe(2);

});
test.each([
  [18, 'par1'],
  [19, 'par1'],
  [20, 'par2'],
  [21, 'par2'],
  [22, 'par2'],
])('Signature help procedure instantiation character %d expecting parameter %s', async (characterReal, parameter) => {
  const linter = await prepare('entity.vhd');
  const help = signatureHelp(linter, Position.create(21, characterReal - 1));
  expect(help?.signatures[0].parameters?.[help?.signatures[0].activeParameter ?? -1]?.label).toBe(parameter);
});
test('Signature help with component', async () => {
  const linter = await prepare('component.vhd');
  expect(signatureHelp(linter, Position.create(21, 42))?.signatures[0].label).toBe('port1, port2, port3');
});