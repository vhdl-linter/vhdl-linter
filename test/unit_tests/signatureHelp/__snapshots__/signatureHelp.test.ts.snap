// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signature help snapshot generic 1`] = `
{
  "signatures": [
    {
      "activeParameter": 0,
      "label": "GENERIC_A, GENERIC_B",
      "parameters": [
        {
          "documentation": {
            "kind": "markdown",
            "value": "\`\`\`vhdl
GENERIC_A : integer
\`\`\`",
          },
          "label": "GENERIC_A",
        },
        {
          "documentation": {
            "kind": "markdown",
            "value": "\`\`\`vhdl
GENERIC_B : integer
\`\`\`",
          },
          "label": "GENERIC_B",
        },
      ],
    },
  ],
}
`;
